// @sipheron/vdr-core/dist/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SipHeron = void 0;
const batch_1 = require("../anchor/batch");
const config_1 = require("./config");
const http_1 = require("./http");
const hash_1 = require("../hash");
const errors_1 = require("../errors");
const cache_1 = require("../verify/cache");
/**
 * Main SipHeron VDR client.
 *
 * @example
 * const sipheron = new SipHeron({ apiKey: 'your-key', network: 'devnet' })
 * const anchor = await sipheron.anchor({ file, name: 'Contract v1' })
 * const result = await sipheron.verify({ file })
 */
class SipHeron {
    constructor(config) {
        this.anchors = {
            revoke: async (anchorId, options) => {
                if (!this.config.apiKey) {
                    throw new errors_1.AuthenticationError('apiKey is required to revoke anchors.');
                }
                await (0, http_1.withRetry)(() => this.http.post(`/api/anchors/${anchorId}/revoke`, options), this.config.retries);
            },
            getVersionChain: async (anchorId) => {
                const endpoint = !this.config.apiKey
                    ? `/api/playground/chain/${anchorId}`
                    : `/api/hashes/${anchorId}/chain`;
                const response = await (0, http_1.withRetry)(() => this.http.get(endpoint), this.config.retries);
                return response.data.chain.map((c) => this._mapAnchorResponse(c));
            }
        };
        this.config = (0, config_1.resolveConfig)(config);
        this.http = (0, http_1.createHttpClient)(this.config);
        if (this.config.cache) {
            this.verifyCache = new cache_1.VerificationCache(this.config.cache);
        }
    }
    /**
     * Anchors a document's SHA-256 fingerprint permanently to
     * the Solana blockchain. The document itself is never
     * transmitted — only its cryptographic fingerprint.
     *
     * @param options - Anchor configuration
     * @param options.file - Document as Buffer. Hashing occurs
     *   locally before any network call. Provide either `file`
     *   or `hash`, not both.
     * @param options.hash - Pre-computed SHA-256 hash (64 hex chars).
     *   Use this when you've already hashed the document yourself.
     * @param options.name - Human-readable document name for
     *   display in the dashboard. Not stored on-chain.
     * @param options.metadata - Arbitrary key-value pairs stored
     *   in SipHeron's database alongside the anchor record.
     *   Not stored on-chain. Max 20 keys, 500 chars per value.
     *
     * @returns Promise resolving to AnchorResult containing the
     *   Solana transaction signature, block number, timestamp,
     *   and permanent verification URL.
     *
     * @throws {AuthenticationError} If API key is invalid or missing
     * @throws {ValidationError} If hash format is invalid
     * @throws {RateLimitError} If monthly quota is exceeded
     * @throws {NetworkError} If Solana RPC is unreachable after retries
     *
     * @example
     * // Anchor from file buffer
     * const file = readFileSync('./contract.pdf')
     * const anchor = await client.anchor({ file, name: 'Q4 Agreement' })
     * console.log(anchor.verificationUrl)
     * // https://verify.sipheron.com/a3f4b2c1...
     *
     * @example
     * // Anchor from pre-computed hash
     * const hash = await hashDocument(fileBuffer)
     * const anchor = await client.anchor({ hash })
     *
     * @see {@link https://docs.sipheron.com/api/anchor}
     * @since 0.1.0
     */
    async anchor(options) {
        // Validate input
        if (!options.file && !options.hash) {
            throw new errors_1.ValidationError('Provide either file (Buffer) or hash (string). Neither was provided.');
        }
        if (options.file && options.hash) {
            throw new errors_1.ValidationError('Provide either file or hash — not both.');
        }
        // Compute hash locally if file provided
        let hash;
        const algorithm = options.hashAlgorithm ?? 'sha256';
        if (options.file) {
            hash = await (0, hash_1.hashDocument)(options.file, { algorithm });
        }
        else {
            hash = (0, hash_1.normalizeHash)(options.hash);
            if (!(0, hash_1.isValidHash)(hash, algorithm)) {
                const info = (0, hash_1.getAlgorithmInfo)(algorithm);
                throw new errors_1.ValidationError(`Invalid hash format for ${info.label}. Must be a ${info.outputLength}-character lowercase hex string.`);
            }
        }
        // Build request headers
        const headers = {};
        if (options.idempotencyKey) {
            headers['Idempotency-Key'] = options.idempotencyKey;
        }
        if (!this.config.apiKey && this.config.network !== 'devnet') {
            throw new errors_1.AuthenticationError('apiKey is required to anchor on mainnet.');
        }
        const endpoint = !this.config.apiKey ? '/api/playground/anchor' : '/api/hashes';
        // Post to API
        const response = await (0, http_1.withRetry)(() => this.http.post(endpoint, {
            hash,
            hashAlgorithm: algorithm,
            filename: options.name || options.metadata?.name || null,
            metadata: options.name || options.metadata?.name || null,
            previousAnchorId: options.previousAnchorId || null,
            ...(options.metadata && { tags: Object.keys(options.metadata) }),
        }, { headers }), this.config.retries);
        return this._mapAnchorResponse(response.data);
    }
    /**
     * Anchors multiple documents in a single batch request efficiently.
     * Instead of waiting for individual Solana transactions, this function
     * submits hashes to the SipHeron queue where they are processed via
     * bulk on-chain transactions or parallelized natively.
     *
     * @param options - Batch anchor configuration
     * @param options.documents - Array of up to 500 documents. Each document
     *   object must contain either `file` (Buffer) or `hash` (string).
     * @param options.stopOnError - If true, the batch fails entirely on the
     *   first error. If false, operations on successful documents continue.
     *
     * @returns Promise resolving to a BatchAnchorResult containing summary stats
     *   and an array of discrete results/errors for each input.
     *
     * @throws {AuthenticationError} If API key is missing (devnet playground
     *   does not support batch operations).
     * @throws {ValidationError} If the batch size exceeds 500 or is empty.
     *
     * @example
     * const results = await client.anchorBatch({
     *   documents: [
     *     { file: buffer1, name: 'Doc 1' },
     *     { file: buffer2, name: 'Doc 2' }
     *   ]
     * })
     * console.log(`Anchored ${results.successful} out of ${results.total}`)
     *
     * @see {@link https://docs.sipheron.com/api/anchor-batch}
     * @since 0.1.0
     */
    async anchorBatch(options) {
        if (!this.config.apiKey) {
            throw new errors_1.AuthenticationError('apiKey is required for batch anchoring. Use single anchor() for devnet playground.');
        }
        return (0, batch_1.anchorBatch)(this, options);
    }
    /**
     * Verifies the absolute authenticity of a document against the
     * immutable Solana blockchain.
     *
     * When providing a `file` buffer, the document is locally hashed,
     * guaranteeing zero-knowledge verification — document bytes are
     * never transmitted over the network.
     *
     * @param options - Verification configuration
     * @param options.file - Document as Buffer to be hashed locally.
     *   Cannot be provided simultaneously with `hash`.
     * @param options.hash - Pre-computed SHA-256 hash string to verify.
     *   Cannot be provided simultaneously with `file`.
     *
     * @returns Promise resolving to VerificationResult.
     *   `authentic: true` implies a perfect cryptograhic match exists on-chain.
     *   `authentic: false` implies the document has been altered or never existed.
     *
     * @throws {ValidationError} If inputs are malformed or missing both params
     * @throws {AnchorRevokedError} If the document anchor was computationally revoked
     * @throws {NetworkError} If SIPHERON verification nodes are completely unavailable
     *
     * @example
     * const file = readFileSync('./received_contract.pdf')
     * const result = await client.verify({ file })
     *
     * if (result.authentic) {
     *   console.log('Valid! Anchored at:', result.timestamp)
     * } else {
     *   console.error('TAMPERED OR UNKNOWN DOCUMENT')
     * }
     *
     * @see {@link https://docs.sipheron.com/api/verify}
     * @since 0.1.0
     */
    async verify(options) {
        if (!options.file && !options.hash) {
            throw new errors_1.ValidationError('Provide either file (Buffer) or hash (string). Neither was provided.');
        }
        // Compute hash locally if file provided
        let hash;
        const algorithm = options.hashAlgorithm ?? 'sha256';
        if (options.file) {
            hash = await (0, hash_1.hashDocument)(options.file, { algorithm });
        }
        else {
            hash = (0, hash_1.normalizeHash)(options.hash);
            if (!(0, hash_1.isValidHash)(hash, algorithm)) {
                const info = (0, hash_1.getAlgorithmInfo)(algorithm);
                throw new errors_1.ValidationError(`Invalid hash format for ${info.label}. Must be a ${info.outputLength}-character hex string.`);
            }
        }
        const endpoint = !this.config.apiKey ? '/api/playground/verify' : '/api/verify';
        // Check cache first
        if (!options.noCache && this.verifyCache) {
            const cached = this.verifyCache.get(hash);
            if (cached)
                return cached;
        }
        const response = await (0, http_1.withRetry)(() => this.http.post(endpoint, { hash, hashAlgorithm: algorithm }), this.config.retries);
        const data = response.data;
        // Handle revoked
        if (data.status === 'REVOKED' || data.anchor?.revocation) {
            // Return revocation info instead of throwing
            const anchorData = data.anchor || {};
            const revocationData = anchorData.revocation || {
                anchorId: anchorData.id || '',
                revokedAt: anchorData.revokedAt || data.verified_at,
                revokedBy: anchorData.revokedBy || 'system',
                reason: anchorData.revocationReason || 'other',
                note: anchorData.revocationNote,
                supersededByAnchorId: anchorData.supersededByAnchorId
            };
            const result = {
                authentic: data.authentic === true || true, // Treat as authentic but superseded
                status: 'revoked',
                hash,
                verifiedAt: data.verified_at,
                anchor: data.anchor ? this._mapAnchorResponse(data) : undefined,
                anchoredHash: data.anchor?.hash,
                revocation: revocationData
            };
            if (this.verifyCache && !options.noCache) {
                this.verifyCache.set(hash, result);
            }
            return result;
        }
        const result = {
            authentic: data.authentic === true,
            status: this._mapVerifyStatus(data.status),
            hash,
            verifiedAt: data.verified_at,
            anchor: data.anchor ? this._mapAnchorResponse(data) : undefined,
            anchoredHash: data.anchor?.hash,
        };
        // Store in cache
        if (this.verifyCache) {
            this.verifyCache.set(hash, result);
        }
        return result;
    }
    /**
     * Get the current status of an anchor by hash.
     *
     * @param hash - 64-character SHA-256 hash
     * @returns AnchorResult with current status
     */
    async getStatus(hash) {
        const normalized = (0, hash_1.normalizeHash)(hash);
        if (!(0, hash_1.isValidHash)(normalized)) {
            throw new errors_1.ValidationError('Invalid hash format.');
        }
        const endpoint = !this.config.apiKey
            ? `/api/hashes/public/${normalized}`
            : `/api/hashes/${normalized}/status`;
        const response = await (0, http_1.withRetry)(() => this.http.get(endpoint), this.config.retries);
        return this._mapAnchorResponse(response.data);
    }
    /**
     * List anchors for the authenticated organization.
     *
     * @param options - Filtering and pagination options
     */
    async list(options) {
        if (!this.config.apiKey) {
            throw new errors_1.AuthenticationError('apiKey is required to list hashes.');
        }
        const params = new URLSearchParams();
        if (options?.page)
            params.set('page', String(options.page));
        if (options?.limit)
            params.set('limit', String(options.limit));
        if (options?.status)
            params.set('status', options.status);
        if (options?.search)
            params.set('search', options.search);
        const response = await (0, http_1.withRetry)(() => this.http.get(`/api/hashes?${params}`), this.config.retries);
        const data = response.data;
        const records = (data.records || data.hashes || []).map((r) => this._mapAnchorResponse({ anchor: r }));
        return {
            records,
            total: data.total || 0,
            page: data.page || 1,
            pages: data.pages || 1,
        };
    }
    /**
     * Verify a pre-computed hash without providing the document.
     * Use this when you already have the SHA-256 hash.
     *
     * @param hash - 64-character SHA-256 hash
     * @returns VerificationResult
     */
    async verifyHash(hash) {
        return this.verify({ hash });
    }
    // ── Private helpers ──
    _mapAnchorResponse(data) {
        const anchor = (data.anchor || data.record || data);
        const blockchain = (data.blockchain || {});
        const hash = String(anchor.hash || '');
        const txSig = String(anchor.txSignature || blockchain.txSignature || '');
        return {
            id: String(anchor.id || ''),
            hash,
            hashAlgorithm: anchor.hashAlgorithm || 'sha256',
            transactionSignature: txSig,
            blockNumber: Number(anchor.blockNumber || blockchain.blockNumber || 0),
            timestamp: String(anchor.blockTimestamp || anchor.createdAt || anchor.registeredAt || ''),
            verificationUrl: (0, config_1.buildVerifyUrl)(hash),
            explorerUrl: txSig
                ? (0, config_1.buildExplorerUrl)(txSig, this.config.network)
                : '',
            status: this._mapStatus(String(anchor.status || 'pending')),
            name: anchor.metadata ? String(anchor.metadata) : undefined,
            contractAddress: this.config.contractAddress,
            network: this.config.network,
        };
    }
    _mapStatus(status) {
        const s = status.toLowerCase();
        if (s === 'confirmed')
            return 'confirmed';
        if (s === 'failed')
            return 'failed';
        if (s === 'revoked')
            return 'revoked';
        return 'pending';
    }
    _mapVerifyStatus(status) {
        const s = (status || '').toLowerCase();
        if (s === 'confirmed')
            return 'authentic';
        if (s === 'revoked')
            return 'revoked';
        if (s === 'not_found')
            return 'not_found';
        if (s === 'pending')
            return 'pending';
        return 'not_found';
    }
}
exports.SipHeron = SipHeron;
