// @sipheron/vdr-core/dist/client/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnchorMonitor = exports.DEFAULT_RPC_NODES = exports.RPCPool = exports.DEFAULT_RETRY_CONFIG = exports.withRetry = exports.DEFAULTS = exports.buildVerifyUrl = exports.buildExplorerUrl = exports.resolveConfig = exports.SipHeron = void 0;
/**
 * @module client
 *
 * @description
 * Public barrel for the `client` sub-system.
 *
 * Consumers of `@sipheron/vdr-core` should import through the root
 * `index.ts` entry point rather than this barrel directly. This file
 * exists to give internal modules a clean relative import path.
 *
 * ### Exported symbols
 * - `SipHeron`          — the main hosted-platform client class.
 * - `resolveConfig`     — merges user config with defaults.
 * - `buildExplorerUrl`  — builds a Solana Explorer transaction URL.
 * - `buildVerifyUrl`    — builds the public SipHeron verification page URL.
 * - `DEFAULTS`          — the raw default config constants object.
 * - `ResolvedConfig`    — (type) the fully resolved config interface.
 */
var client_1 = require("./client");
Object.defineProperty(exports, "SipHeron", { enumerable: true, get: function () { return client_1.SipHeron; } });
var config_1 = require("./config");
Object.defineProperty(exports, "resolveConfig", { enumerable: true, get: function () { return config_1.resolveConfig; } });
Object.defineProperty(exports, "buildExplorerUrl", { enumerable: true, get: function () { return config_1.buildExplorerUrl; } });
Object.defineProperty(exports, "buildVerifyUrl", { enumerable: true, get: function () { return config_1.buildVerifyUrl; } });
Object.defineProperty(exports, "DEFAULTS", { enumerable: true, get: function () { return config_1.DEFAULTS; } });
var retry_1 = require("./retry");
Object.defineProperty(exports, "withRetry", { enumerable: true, get: function () { return retry_1.withRetry; } });
Object.defineProperty(exports, "DEFAULT_RETRY_CONFIG", { enumerable: true, get: function () { return retry_1.DEFAULT_RETRY_CONFIG; } });
var rpc_pool_1 = require("./rpc-pool");
Object.defineProperty(exports, "RPCPool", { enumerable: true, get: function () { return rpc_pool_1.RPCPool; } });
Object.defineProperty(exports, "DEFAULT_RPC_NODES", { enumerable: true, get: function () { return rpc_pool_1.DEFAULT_RPC_NODES; } });
var monitor_1 = require("./monitor");
Object.defineProperty(exports, "AnchorMonitor", { enumerable: true, get: function () { return monitor_1.AnchorMonitor; } });
