// @sipheron/vdr-core/dist/client/monitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnchorMonitor = void 0;
const events_1 = require("events");
class AnchorMonitor extends events_1.EventEmitter {
    constructor(client, pollIntervalMs = 2000) {
        super();
        this.client = client;
        this.pollIntervalMs = pollIntervalMs;
        this.watched = new Map();
    }
    /**
     * Watch an anchor until it confirms or fails.
     * Emits 'confirmation' on each new confirmation,
     * 'confirmed' when fully confirmed, 'failed' on error.
     */
    watch(anchorId) {
        // Immediate pending push
        this.emit('pending', anchorId);
        const interval = setInterval(async () => {
            try {
                const anchorResult = await this.client.getStatus(anchorId);
                // SipHeron hashes API can incrementally report confirmations. 
                // We gracefully fallback to 1 if the API just transitioned to confirmed.
                const confirmations = anchorResult.confirmations ?? (anchorResult.status === 'confirmed' ? 1 : 0);
                this.emit('confirmation', anchorId, confirmations);
                if (anchorResult.status === 'confirmed') {
                    clearInterval(interval);
                    this.watched.delete(anchorId);
                    this.emit('confirmed', anchorResult);
                }
                else if (anchorResult.status === 'failed' || anchorResult.status === 'revoked') {
                    clearInterval(interval);
                    this.watched.delete(anchorId);
                    this.emit('failed', anchorId, new Error(`Anchor terminated with status: ${anchorResult.status}`));
                }
            }
            catch (error) {
                clearInterval(interval);
                this.watched.delete(anchorId);
                this.emit('failed', anchorId, error);
            }
        }, this.pollIntervalMs);
        this.watched.set(anchorId, interval);
        return this;
    }
    stopWatching(anchorId) {
        const interval = this.watched.get(anchorId);
        if (interval) {
            clearInterval(interval);
            this.watched.delete(anchorId);
        }
    }
    stopAll() {
        this.watched.forEach(interval => clearInterval(interval));
        this.watched.clear();
    }
    // Type overrides for strictly-typed Event Emitter
    on(event, listener) {
        return super.on(event, listener);
    }
    emit(event, ...args) {
        return super.emit(event, ...args);
    }
}
exports.AnchorMonitor = AnchorMonitor;
