// @sipheron/vdr-core/dist/client/retry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RETRY_CONFIG = void 0;
exports.withRetry = withRetry;
exports.DEFAULT_RETRY_CONFIG = {
    maxAttempts: 3,
    baseDelayMs: 1000,
    maxDelayMs: 30000,
    backoffMultiplier: 2,
    retryableErrors: [
        'NETWORK_ERROR',
        'SOLANA_CONNECTION_ERROR',
        'RATE_LIMIT_ERROR',
        'TRANSACTION_TIMEOUT'
    ]
};
async function withRetry(operation, config = {}) {
    const cfg = { ...exports.DEFAULT_RETRY_CONFIG, ...config };
    let lastError;
    for (let attempt = 1; attempt <= cfg.maxAttempts; attempt++) {
        try {
            return await operation();
        }
        catch (error) {
            lastError = error;
            const isRetryable = cfg.retryableErrors.some(code => error.code === code);
            if (!isRetryable || attempt === cfg.maxAttempts) {
                throw error;
            }
            const delay = Math.min(cfg.baseDelayMs * Math.pow(cfg.backoffMultiplier, attempt - 1), cfg.maxDelayMs);
            await new Promise(resolve => setTimeout(resolve, delay));
        }
    }
    throw lastError;
}
