// @sipheron/vdr-core/dist/client/rpc-pool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RPC_NODES = exports.RPCPool = void 0;
/**
 * RPC node pool with automatic failover.
 * Rotates through available nodes when one fails.
 */
class RPCPool {
    constructor(nodes) {
        this.currentIndex = 0;
        this.failures = new Map();
        this.nodes = nodes;
    }
    getCurrent() {
        return this.nodes[this.currentIndex];
    }
    markFailed(node) {
        const failures = (this.failures.get(node) || 0) + 1;
        this.failures.set(node, failures);
        if (failures >= 3) {
            this.rotate();
        }
    }
    markSuccess(node) {
        this.failures.delete(node);
    }
    rotate() {
        this.currentIndex = (this.currentIndex + 1) % this.nodes.length;
    }
}
exports.RPCPool = RPCPool;
exports.DEFAULT_RPC_NODES = {
    devnet: [
        'https://api.devnet.solana.com',
        'https://devnet.helius-rpc.com',
        'https://rpc-devnet.helius.xyz'
    ],
    mainnet: [
        'https://api.mainnet-beta.solana.com',
        'https://rpc.helius.xyz',
        'https://solana-api.projectserum.com'
    ]
};
