// @sipheron/vdr-core/dist/index.js
"use strict";
/**
 * @module @sipheron/vdr-core
 * @version 0.1.18
 * @license Apache-2.0
 *
 * @description
 * The cryptographic engine of SipHeron VDR.
 * Permanently anchor any document's SHA-256 fingerprint to the Solana
 * blockchain and verify authenticity anywhere — forever.
 *
 * This is the **library entry point**. It re-exports every public symbol
 * from the sub-modules so consumers only need a single import path:
 *
 * ```ts
 * import {
 *   SipHeron,           // Hosted platform client (requires API key)
 *   hashDocument,       // SHA-256 hashing — runs 100 % client-side
 *   anchorToSolana,     // Direct on-chain write (no API key needed)
 *   verifyOnChain,      // Direct on-chain read  (no API key needed)
 *   deriveAnchorAddress // PDA derivation math
 * } from '@sipheron/vdr-core'
 * ```
 *
 * ## Two Operating Modes
 *
 * ### DIRECT — no SipHeron account required
 * Uses public Solana RPC nodes exclusively.
 * - `hashDocument / hashFile / hashStream / hashBase64`
 * - `anchorToSolana`  — write to Solana directly with your own Keypair
 * - `verifyOnChain`   — read from Solana directly
 * - `deriveAnchorAddress` — compute the PDA for any hash + owner pair
 * - `verifyLocally`   — constant-time local comparison (no network)
 *
 * ### HOSTED — managed SipHeron platform (API key required for mainnet)
 * Adds managed analytics, PDF certificates, compliance exports, etc.
 * - `new SipHeron({ apiKey, network })`
 * - `sipheron.anchor()` / `sipheron.anchorBatch()`
 * - `sipheron.verify()` / `sipheron.verifyHash()`
 * - `sipheron.list()` / `sipheron.getStatus()`
 *
 * @see {@link https://app.sipheron.com} SipHeron Dashboard
 * @see {@link https://github.com/SipHeron-VDR/vdr-core} Source Repository
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RateLimitError = exports.NetworkError = exports.HashMismatchError = exports.AnchorNotFoundError = exports.AuthenticationError = exports.SipHeronError = exports.anchorToSolana = exports.SOLANA_NETWORKS = exports.CONFIRMATION_DEPTH = exports.HASH_LENGTH = exports.MAX_BATCH_SIZE = exports.PROTOCOL_VERSION = exports.ANCHOR_SEED = exports.SIPHERON_PROGRAM_ID = exports.SIPHERON_CONTRACT = exports.estimateAnchorCost = exports.isValidTxSignature = exports.getExplorerUrl = exports.CLINICAL_TRIAL_SCHEMA = exports.LEGAL_CONTRACT_SCHEMA = exports.validateMetadata = exports.getCertificateProofText = exports.prepareCertificateData = exports.buildCertificateUrl = exports.generateCertificateId = exports.VerificationCache = exports.parseWebhookEvent = exports.verifyWebhookSignature = exports.deriveAnchorAddress = exports.verifyOnChain = exports.verifyLocally = exports.verifyHash = exports.hashAuto = exports.hashDocumentBrowser = exports.getAlgorithmInfo = exports.normalizeHash = exports.isValidHash = exports.hashBase64 = exports.hashStream = exports.hashFileWithProgress = exports.hashFileStream = exports.hashFile = exports.hashDocument = exports.anchorBatch = exports.AnchorMonitor = exports.DEFAULT_RPC_NODES = exports.RPCPool = exports.DEFAULT_RETRY_CONFIG = exports.withRetry = exports.SipHeron = void 0;
exports.TransactionError = exports.SolanaConnectionError = exports.QuotaExceededError = exports.AnchorRevokedError = exports.ValidationError = void 0;
// ── Main client ──
var client_1 = require("./client");
Object.defineProperty(exports, "SipHeron", { enumerable: true, get: function () { return client_1.SipHeron; } });
Object.defineProperty(exports, "withRetry", { enumerable: true, get: function () { return client_1.withRetry; } });
Object.defineProperty(exports, "DEFAULT_RETRY_CONFIG", { enumerable: true, get: function () { return client_1.DEFAULT_RETRY_CONFIG; } });
Object.defineProperty(exports, "RPCPool", { enumerable: true, get: function () { return client_1.RPCPool; } });
Object.defineProperty(exports, "DEFAULT_RPC_NODES", { enumerable: true, get: function () { return client_1.DEFAULT_RPC_NODES; } });
Object.defineProperty(exports, "AnchorMonitor", { enumerable: true, get: function () { return client_1.AnchorMonitor; } });
var anchor_1 = require("./anchor");
Object.defineProperty(exports, "anchorBatch", { enumerable: true, get: function () { return anchor_1.anchorBatch; } });
// ── Standalone functions ──
var hash_1 = require("./hash");
Object.defineProperty(exports, "hashDocument", { enumerable: true, get: function () { return hash_1.hashDocument; } });
Object.defineProperty(exports, "hashFile", { enumerable: true, get: function () { return hash_1.hashFile; } });
Object.defineProperty(exports, "hashFileStream", { enumerable: true, get: function () { return hash_1.hashFileStream; } });
Object.defineProperty(exports, "hashFileWithProgress", { enumerable: true, get: function () { return hash_1.hashFileWithProgress; } });
Object.defineProperty(exports, "hashStream", { enumerable: true, get: function () { return hash_1.hashStream; } });
Object.defineProperty(exports, "hashBase64", { enumerable: true, get: function () { return hash_1.hashBase64; } });
Object.defineProperty(exports, "isValidHash", { enumerable: true, get: function () { return hash_1.isValidHash; } });
Object.defineProperty(exports, "normalizeHash", { enumerable: true, get: function () { return hash_1.normalizeHash; } });
Object.defineProperty(exports, "getAlgorithmInfo", { enumerable: true, get: function () { return hash_1.getAlgorithmInfo; } });
var hash_2 = require("./hash");
Object.defineProperty(exports, "hashDocumentBrowser", { enumerable: true, get: function () { return hash_2.hashDocumentBrowser; } });
Object.defineProperty(exports, "hashAuto", { enumerable: true, get: function () { return hash_2.hashAuto; } });
var verify_1 = require("./verify");
Object.defineProperty(exports, "verifyHash", { enumerable: true, get: function () { return verify_1.verifyHashStandalone; } });
var verify_2 = require("./verify");
Object.defineProperty(exports, "verifyLocally", { enumerable: true, get: function () { return verify_2.verifyLocally; } });
Object.defineProperty(exports, "verifyOnChain", { enumerable: true, get: function () { return verify_2.verifyOnChain; } });
Object.defineProperty(exports, "deriveAnchorAddress", { enumerable: true, get: function () { return verify_2.deriveAnchorAddress; } });
var webhook_1 = require("./webhook");
Object.defineProperty(exports, "verifyWebhookSignature", { enumerable: true, get: function () { return webhook_1.verifyWebhookSignature; } });
Object.defineProperty(exports, "parseWebhookEvent", { enumerable: true, get: function () { return webhook_1.parseWebhookEvent; } });
var cache_1 = require("./verify/cache");
Object.defineProperty(exports, "VerificationCache", { enumerable: true, get: function () { return cache_1.VerificationCache; } });
// ── Certificate utilities ──
var certificate_1 = require("./certificate");
Object.defineProperty(exports, "generateCertificateId", { enumerable: true, get: function () { return certificate_1.generateCertificateId; } });
Object.defineProperty(exports, "buildCertificateUrl", { enumerable: true, get: function () { return certificate_1.buildCertificateUrl; } });
Object.defineProperty(exports, "prepareCertificateData", { enumerable: true, get: function () { return certificate_1.prepareCertificateData; } });
Object.defineProperty(exports, "getCertificateProofText", { enumerable: true, get: function () { return certificate_1.getCertificateProofText; } });
// ── Metadata Schema & Validation ──
var schema_1 = require("./schema");
Object.defineProperty(exports, "validateMetadata", { enumerable: true, get: function () { return schema_1.validateMetadata; } });
Object.defineProperty(exports, "LEGAL_CONTRACT_SCHEMA", { enumerable: true, get: function () { return schema_1.LEGAL_CONTRACT_SCHEMA; } });
Object.defineProperty(exports, "CLINICAL_TRIAL_SCHEMA", { enumerable: true, get: function () { return schema_1.CLINICAL_TRIAL_SCHEMA; } });
// ── Solana utilities ──
var anchor_2 = require("./anchor");
Object.defineProperty(exports, "getExplorerUrl", { enumerable: true, get: function () { return anchor_2.getExplorerUrl; } });
Object.defineProperty(exports, "isValidTxSignature", { enumerable: true, get: function () { return anchor_2.isValidTxSignature; } });
Object.defineProperty(exports, "estimateAnchorCost", { enumerable: true, get: function () { return anchor_2.estimateAnchorCost; } });
Object.defineProperty(exports, "SIPHERON_CONTRACT", { enumerable: true, get: function () { return anchor_2.SIPHERON_CONTRACT; } });
Object.defineProperty(exports, "SIPHERON_PROGRAM_ID", { enumerable: true, get: function () { return anchor_2.SIPHERON_PROGRAM_ID; } });
Object.defineProperty(exports, "ANCHOR_SEED", { enumerable: true, get: function () { return anchor_2.ANCHOR_SEED; } });
Object.defineProperty(exports, "PROTOCOL_VERSION", { enumerable: true, get: function () { return anchor_2.PROTOCOL_VERSION; } });
Object.defineProperty(exports, "MAX_BATCH_SIZE", { enumerable: true, get: function () { return anchor_2.MAX_BATCH_SIZE; } });
Object.defineProperty(exports, "HASH_LENGTH", { enumerable: true, get: function () { return anchor_2.HASH_LENGTH; } });
Object.defineProperty(exports, "CONFIRMATION_DEPTH", { enumerable: true, get: function () { return anchor_2.CONFIRMATION_DEPTH; } });
Object.defineProperty(exports, "SOLANA_NETWORKS", { enumerable: true, get: function () { return anchor_2.SOLANA_NETWORKS; } });
Object.defineProperty(exports, "anchorToSolana", { enumerable: true, get: function () { return anchor_2.anchorToSolana; } });
// ── Errors ──
var errors_1 = require("./errors");
Object.defineProperty(exports, "SipHeronError", { enumerable: true, get: function () { return errors_1.SipHeronError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return errors_1.AuthenticationError; } });
Object.defineProperty(exports, "AnchorNotFoundError", { enumerable: true, get: function () { return errors_1.AnchorNotFoundError; } });
Object.defineProperty(exports, "HashMismatchError", { enumerable: true, get: function () { return errors_1.HashMismatchError; } });
Object.defineProperty(exports, "NetworkError", { enumerable: true, get: function () { return errors_1.NetworkError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return errors_1.RateLimitError; } });
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return errors_1.ValidationError; } });
Object.defineProperty(exports, "AnchorRevokedError", { enumerable: true, get: function () { return errors_1.AnchorRevokedError; } });
Object.defineProperty(exports, "QuotaExceededError", { enumerable: true, get: function () { return errors_1.QuotaExceededError; } });
Object.defineProperty(exports, "SolanaConnectionError", { enumerable: true, get: function () { return errors_1.SolanaConnectionError; } });
Object.defineProperty(exports, "TransactionError", { enumerable: true, get: function () { return errors_1.TransactionError; } });
