// @sipheron/vdr-core/dist/verify/cache.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificationCache = void 0;
const fs = __importStar(require("fs"));
/**
 * Local cache for verification results to prevent redundant network calls.
 *
 * Support in-memory storage with optional LRU-style eviction and
 * filesystem persistence.
 */
class VerificationCache {
    constructor(options = {}) {
        this.cache = new Map();
        this.ttlMs = options.ttlMs ?? 300000;
        this.maxEntries = options.maxEntries ?? 1000;
        this.persistPath = options.persistPath;
        if (this.persistPath) {
            this.loadFromDisk();
        }
    }
    /**
     * Look up a verification result by document hash.
     *
     * @param hash - The SHA-256 fingerprint to look up.
     * @returns The cached VerificationResult if valid, or null if expired/missing.
     */
    get(hash) {
        const entry = this.cache.get(hash);
        if (!entry)
            return null;
        const now = Date.now();
        if (now - entry.cachedAt > entry.ttlMs) {
            this.cache.delete(hash);
            return null;
        }
        return {
            ...entry.result,
            fromCache: true,
            cachedTimestamp: entry.cachedAt
        };
    }
    /**
     * Store a verification result in the cache.
     *
     * @param hash - The SHA-256 fingerprint of the document.
     * @param result - The fresh verification outcome to cache.
     */
    set(hash, result) {
        if (this.cache.size >= this.maxEntries && !this.cache.has(hash)) {
            // Basic LRU: Evict the oldest entry based on cachedAt
            let oldestHash = null;
            let oldestTime = Infinity;
            for (const [h, entry] of this.cache.entries()) {
                if (entry.cachedAt < oldestTime) {
                    oldestTime = entry.cachedAt;
                    oldestHash = h;
                }
            }
            if (oldestHash) {
                this.cache.delete(oldestHash);
            }
        }
        this.cache.set(hash, {
            result,
            cachedAt: Date.now(),
            ttlMs: this.ttlMs
        });
        if (this.persistPath) {
            this.saveToDisk();
        }
    }
    /**
     * Remove a specific hash from the cache.
     */
    invalidate(hash) {
        this.cache.delete(hash);
        if (this.persistPath) {
            this.saveToDisk();
        }
    }
    /**
     * Wipe the entire cache.
     */
    clear() {
        this.cache.clear();
        if (this.persistPath) {
            this.saveToDisk();
        }
    }
    loadFromDisk() {
        if (!this.persistPath)
            return;
        try {
            if (!fs.existsSync(this.persistPath))
                return;
            const raw = fs.readFileSync(this.persistPath, 'utf-8');
            const data = JSON.parse(raw);
            // Data is stored as an object, reconstruct Map
            for (const [key, value] of Object.entries(data)) {
                this.cache.set(key, value);
            }
        }
        catch {
            // Silently fail on corruption
        }
    }
    saveToDisk() {
        if (!this.persistPath)
            return;
        try {
            const data = Object.fromEntries(this.cache);
            fs.writeFileSync(this.persistPath, JSON.stringify(data, null, 2), 'utf-8');
        }
        catch {
            // Silently fail on write permissions/disk issues
        }
    }
}
exports.VerificationCache = VerificationCache;
