// @sipheron/vdr-core/dist/anchor/anchor.js
"use strict";
/**
 * @module anchor/anchor
 *
 * @description
 * Core anchoring helpers used internally by the `SipHeron` client.
 * These are the pure utility functions that prepare, validate, and map
 * anchor data — they contain no HTTP logic themselves.
 *
 * ## Contents
 * - `prepareAnchor(options)`  — Normalises `AnchorOptions` into a `{ hash, metadata }` object.
 *   Hashes the file locally if a Buffer is provided; validates the format if a raw hash string
 *   is provided. The document bytes are **never** transmitted.
 * - `mapToAnchorResult(data)` — Maps a raw API response object to the typed `AnchorResult`.
 * - `hashDocument`            — Re-exported for convenience in anchor workflows.
 *
 * @internal
 * This module is not part of the public API surface — consumers should use
 * the `SipHeron` client class or the standalone functions from `@sipheron/vdr-core` directly.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hashDocument = void 0;
exports.prepareAnchor = prepareAnchor;
exports.mapToAnchorResult = mapToAnchorResult;
const hash_1 = require("../hash");
Object.defineProperty(exports, "hashDocument", { enumerable: true, get: function () { return hash_1.hashDocument; } });
const errors_1 = require("../errors");
const config_1 = require("../client/config");
/**
 * Prepare anchor data from AnchorOptions.
 * Computes hash if file is provided, validates hash if string is provided.
 * Returns the normalized hash ready for API submission.
 */
async function prepareAnchor(options) {
    if (!options.file && !options.hash) {
        throw new errors_1.ValidationError('Provide either file (Buffer) or hash (string).');
    }
    let hash;
    if (options.file) {
        hash = await (0, hash_1.hashDocument)(options.file);
    }
    else {
        hash = (0, hash_1.normalizeHash)(options.hash);
        if (!(0, hash_1.isValidHash)(hash)) {
            throw new errors_1.ValidationError('Invalid hash format. Must be 64 hex characters (SHA-256).');
        }
    }
    return {
        hash,
        metadata: options.name || null,
    };
}
/**
 * Map raw API response to AnchorResult.
 */
function mapToAnchorResult(data, network = 'devnet') {
    const record = (data.record || data);
    const hash = String(record.hash || '');
    const txSig = String(record.txSignature || '');
    return {
        id: String(record.id || ''),
        hash,
        transactionSignature: txSig,
        blockNumber: Number(record.blockNumber || 0),
        timestamp: String(record.blockTimestamp || record.createdAt || ''),
        verificationUrl: (0, config_1.buildVerifyUrl)(hash),
        explorerUrl: txSig ? (0, config_1.buildExplorerUrl)(txSig, network) : '',
        status: ['confirmed', 'failed', 'revoked', 'pending']
            .find(s => s === record.status?.toString().toLowerCase()) || 'pending',
        name: record.metadata ? String(record.metadata) : undefined,
        contractAddress: '6ecWPUK87zxwZP2pARJ75wbpCka92mYSGP1szrJxzAwo',
        network,
    };
}
