// @sipheron/vdr-core/dist/anchor/batch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.anchorBatch = anchorBatch;
async function anchorBatch(client, options) {
    const { documents, concurrency = 5, onProgress, continueOnError = true, delayBetweenBatchesMs = 100 } = options;
    const startTime = Date.now();
    const results = [];
    let succeeded = 0;
    let failed = 0;
    // Process in chunks respecting concurrency limit
    for (let i = 0; i < documents.length; i += concurrency) {
        const chunk = documents.slice(i, i + concurrency);
        const chunkResults = await Promise.allSettled(chunk.map(async (doc, chunkIndex) => {
            const index = i + chunkIndex;
            try {
                const anchor = await client.anchor({
                    ...doc,
                    hashAlgorithm: doc.hashAlgorithm || options.hashAlgorithm
                });
                return { index, success: true, anchor };
            }
            catch (error) {
                if (!continueOnError)
                    throw error;
                return {
                    index,
                    success: false,
                    error: error.message
                };
            }
        }));
        for (let c = 0; c < chunkResults.length; c++) {
            const result = chunkResults[c];
            const index = i + c;
            let item;
            if (result.status === 'fulfilled') {
                item = result.value;
            }
            else {
                item = { index, success: false, error: result.reason?.message };
            }
            results.push(item);
            if (item.success)
                succeeded++;
            else
                failed++;
            onProgress?.(results.length, documents.length, item);
        }
        // Respect rate limits between chunks
        if (i + concurrency < documents.length) {
            await new Promise(r => setTimeout(r, delayBetweenBatchesMs));
        }
    }
    return {
        total: documents.length,
        succeeded,
        failed,
        results: results.sort((a, b) => a.index - b.index),
        durationMs: Date.now() - startTime
    };
}
