// @sipheron/vdr-core/dist/anchor/direct.js
"use strict";
/**
 * @module anchor/direct
 *
 * @description
 * **Direct on-chain anchoring** — no SipHeron account or API key required.
 *
 * `anchorToSolana()` writes a document's SHA-256 hash directly to the Solana
 * blockchain by invoking the SipHeron smart contract via JSON-RPC. The caller
 * must supply their own funded Solana `Keypair`; the library constructs and
 * broadcasts the transaction autonomously.
 *
 * ### How it works
 * 1. Hash the document locally (or accept a pre-computed hash).
 * 2. Derive two Program Derived Addresses (PDAs):
 *    - `protocol_config` — reads the on-chain treasury address.
 *    - `hash_record`     — the account that stores this document's fingerprint.
 * 3. Call `program.methods.registerHash(hashArray, metadata, expiry)`.
 * 4. Return the transaction signature, Solana Explorer URL, and PDA address.
 *
 * ### Custom program support
 * Pass `programId` (base58 string) in the options to target your own
 * deployed fork of the VDR smart contract instead of the default SipHeron one.
 *
 * @example
 * ```ts
 * import { anchorToSolana } from '@sipheron/vdr-core'
 * import { Keypair } from '@solana/web3.js'
 * import fs from 'fs'
 *
 * const keypair = Keypair.fromSecretKey(mySecretKeyBytes)
 * const buffer  = fs.readFileSync('./contract.pdf')
 *
 * const result = await anchorToSolana({ buffer, keypair, network: 'mainnet' })
 * console.log('PDA:', result.pda)
 * console.log('Explorer:', result.explorerUrl)
 * ```
 *
 * @see {@link DirectAnchorOptions} for the full options reference.
 * @see {@link DirectAnchorResult}  for the return value shape.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.anchorToSolana = anchorToSolana;
const anchor = __importStar(require("@coral-xyz/anchor"));
const web3_js_1 = require("@solana/web3.js");
const hash_1 = require("../hash");
const solana_1 = require("./solana");
const idl_json_1 = __importDefault(require("./idl.json"));
const errors_1 = require("../errors");
/**
 * Directly anchors a document hash to the Solana blockchain.
 * Does not require a SipHeron account or API key.
 *
 * Requirements:
 * 1. A Solana Keypair with sufficient SOL (~0.000005 SOL per anchor).
 *
 * @param options DirectAnchorOptions
 * @returns Complete transaction details including the PDA address.
 */
async function anchorToSolana(options) {
    const { keypair, network, metadata = 'Direct Anchor via vdr-core', rpcUrl, programId: customProgramId } = options;
    if (!options.hash && !options.buffer) {
        throw new errors_1.ValidationError('Must provide either a hash or file buffer');
    }
    const hashString = options.buffer
        ? await (0, hash_1.hashDocument)(options.buffer)
        : options.hash.toLowerCase();
    const hashBuffer = Buffer.from(hashString, 'hex');
    const hashArray = Array.from(hashBuffer);
    const connectionUrl = rpcUrl || solana_1.SOLANA_NETWORKS[network];
    const connection = new web3_js_1.Connection(connectionUrl, 'confirmed');
    const programId = customProgramId
        ? new web3_js_1.PublicKey(customProgramId)
        : solana_1.SIPHERON_PROGRAM_ID[network];
    const provider = new anchor.AnchorProvider(connection, new anchor.Wallet(keypair), { preflightCommitment: 'confirmed' });
    const program = new anchor.Program(idl_json_1.default, provider);
    // Derive PDAs
    const [protocolConfigPda] = web3_js_1.PublicKey.findProgramAddressSync([Buffer.from('protocol_config')], programId);
    const [hashPda] = web3_js_1.PublicKey.findProgramAddressSync([Buffer.from('hash_record'), hashBuffer, keypair.publicKey.toBuffer()], programId);
    try {
        // We need to fetch the treasury public key from the protocol config
        // Raw fetch bypasses IDL size mismatches between devnet/mainnet versions during upgrades
        const accountInfo = await connection.getAccountInfo(protocolConfigPda);
        if (!accountInfo)
            throw new Error('Protocol config not found on-chain');
        // discriminator (8) + admin (32) + fee (8) = 48 -> treasury starts at 48
        const treasury = new web3_js_1.PublicKey(accountInfo.data.slice(48, 80));
        const txSig = await program.methods
            .registerHash(hashArray, metadata, new anchor.BN(0)) // 0 means no expiry
            .accounts({
            hashRecord: hashPda,
            protocolConfig: protocolConfigPda,
            treasury,
            organization: null, // Only used in managed platform anchors
            owner: keypair.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        })
            .signers([keypair])
            .rpc();
        return {
            hash: hashString,
            transactionSignature: txSig,
            network,
            explorerUrl: (0, solana_1.getExplorerUrl)(txSig, network),
            cost: (0, solana_1.estimateAnchorCost)().lamports,
            pda: hashPda.toBase58()
        };
    }
    catch (err) {
        if (err.message && err.message.includes('fetch')) {
            throw new errors_1.SolanaConnectionError(`Failed to connect to Solana or fetch protocol config: ${err.message}`);
        }
        throw new errors_1.TransactionError(`Failed to broadcast transaction: ${err.message}`);
    }
}
