// @sipheron/vdr-core/dist/anchor/solana.js
"use strict";
/**
 * @module anchor/solana
 *
 * @description
 * Solana-specific constants and utility functions for the VDR SDK.
 *
 * This module is a **pure utility layer** — it does not send any transactions
 * or manage private keys. All signing happens in `anchor/direct.ts` using the
 * caller-supplied `Keypair`.
 *
 * ## Constants
 * | Name                     | Value / Description                                        |
 * |--------------------------|------------------------------------------------------------|
 * | `SOLANA_NETWORKS`        | Public RPC endpoints for devnet and mainnet.               |
 * | `SIPHERON_PROGRAM_ID`    | `PublicKey` map keyed by network for the VDR contract.     |
 * | `SIPHERON_CONTRACT`      | Base58 string of the deployed program address.             |
 * | `ANCHOR_SEED`            | PDA seed string prefix (`'anchor'`).                       |
 * | `PROTOCOL_VERSION`       | Current on-chain protocol version integer.                 |
 * | `MAX_BATCH_SIZE`         | Maximum documents per batch anchor call (500).             |
 * | `HASH_LENGTH`            | Required SHA-256 hex digest length (64 chars).             |
 * | `CONFIRMATION_DEPTH`     | Enterprise confirmation block depth (32).                  |
 * | `EXPLORER_URLS`          | Helper functions for Solana Explorer deep-links.           |
 *
 * ## Functions
 * - `getExplorerUrl(txSig, network)` — Build a Solana Explorer URL for a transaction.
 * - `isValidTxSignature(sig)`        — Validate base58 tx signature format.
 * - `estimateAnchorCost()`           — Return estimated lamport cost per anchor.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EXPLORER_URLS = exports.SIPHERON_CONTRACT = exports.CONFIRMATION_DEPTH = exports.HASH_LENGTH = exports.MAX_BATCH_SIZE = exports.PROTOCOL_VERSION = exports.ANCHOR_SEED = exports.SIPHERON_PROGRAM_ID = exports.SIPHERON_CONTRACT_BASE58 = exports.SOLANA_NETWORKS = void 0;
exports.getExplorerUrl = getExplorerUrl;
exports.isValidTxSignature = isValidTxSignature;
exports.estimateAnchorCost = estimateAnchorCost;
const web3_js_1 = require("@solana/web3.js");
exports.SOLANA_NETWORKS = {
    devnet: 'https://api.devnet.solana.com',
    mainnet: 'https://api.mainnet-beta.solana.com',
};
exports.SIPHERON_CONTRACT_BASE58 = '6ecWPUK87zxwZP2pARJ75wbpCka92mYSGP1szrJxzAwo';
exports.SIPHERON_PROGRAM_ID = {
    devnet: new web3_js_1.PublicKey(exports.SIPHERON_CONTRACT_BASE58),
    mainnet: new web3_js_1.PublicKey(exports.SIPHERON_CONTRACT_BASE58)
};
exports.ANCHOR_SEED = 'anchor';
exports.PROTOCOL_VERSION = 1;
exports.MAX_BATCH_SIZE = 500;
exports.HASH_LENGTH = 64;
exports.CONFIRMATION_DEPTH = 32;
exports.SIPHERON_CONTRACT = exports.SIPHERON_CONTRACT_BASE58;
exports.EXPLORER_URLS = {
    devnet: {
        tx: (sig) => `https://explorer.solana.com/tx/${sig}?cluster=devnet`,
        address: (addr) => `https://explorer.solana.com/address/${addr}?cluster=devnet`,
        block: (slot) => `https://explorer.solana.com/block/${slot}?cluster=devnet`,
    },
    mainnet: {
        tx: (sig) => `https://explorer.solana.com/tx/${sig}`,
        address: (addr) => `https://explorer.solana.com/address/${addr}`,
        block: (slot) => `https://explorer.solana.com/block/${slot}`,
    },
};
/**
 * Build a Solana Explorer URL for a transaction.
 */
function getExplorerUrl(txSignature, network = 'devnet') {
    return exports.EXPLORER_URLS[network].tx(txSignature);
}
/**
 * Validate a Solana transaction signature format.
 * Signatures are base58-encoded 64-byte values — typically 87-88 chars.
 */
function isValidTxSignature(sig) {
    return typeof sig === 'string' &&
        sig.length >= 43 &&
        sig.length <= 128 &&
        /^[1-9A-HJ-NP-Za-km-z]+$/.test(sig);
}
/**
 * Estimate the approximate SOL cost of anchoring a document.
 * Based on average Solana transaction fees as of 2026.
 */
function estimateAnchorCost() {
    const lamports = 5000; // ~0.000005 SOL per transaction
    return {
        lamports,
        sol: lamports / 1000000000,
        usdApprox: '< $0.01',
    };
}
