// @sipheron/vdr-core/dist/hash/algorithms.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAlgorithmInfo = getAlgorithmInfo;
exports.hashDocument = hashDocument;
exports.hashFile = hashFile;
exports.hashFileStream = hashFileStream;
exports.hashFileWithProgress = hashFileWithProgress;
exports.hashBase64 = hashBase64;
exports.hashStream = hashStream;
exports.isValidHash = isValidHash;
exports.normalizeHash = normalizeHash;
/**
 * @module hash/algorithms
 *
 * @description
 * Multi-algorithm hashing engine supporting SHA-256, SHA-512, BLAKE3, and MD5.
 *
 * All hashing runs **100% locally**. SipHeron VDR provides flexibility for
 * enterprise integration by supporting multiple standards while maintaining
 * cryptographic integrity.
 *
 * ## Performance note
 * - **BLAKE3** is the fastest for large document streams but requires an
 *   optional native dependency.
 * - **SHA-256** is the industry standard and works natively everywhere.
 * - **SHA-512** provides maximum collision resistance.
 */
const crypto_1 = require("crypto");
const errors_1 = require("../errors");
/**
 * Returns metadata about a supported hashing algorithm.
 */
function getAlgorithmInfo(algorithm) {
    const info = {
        sha256: { bits: 256, outputLength: 64, recommended: true, label: 'SHA-256' },
        sha512: { bits: 512, outputLength: 128, recommended: true, label: 'SHA-512' },
        blake3: { bits: 256, outputLength: 64, recommended: true, fast: true, label: 'BLAKE3' },
        md5: { bits: 128, outputLength: 32, recommended: false, label: 'MD5',
            warning: 'MD5 is cryptographically broken. Use only for legacy compatibility.' }
    };
    return info[algorithm];
}
/**
 * Internal helper to dynamically import BLAKE3 if available.
 */
async function importBlake3() {
    try {
        // @ts-ignore - optional dependency
        return await Promise.resolve().then(() => __importStar(require('blake3')));
    }
    catch (e) {
        throw new Error('BLAKE3 support requires the "blake3" package to be installed. ' +
            'Run: npm install blake3');
    }
}
/**
 * Compute the hash of a document buffer using the specified algorithm.
 *
 * @param file - Document as Buffer
 * @param options - Hashing configuration (algorithm, encoding)
 * @returns Lowercase hex or base64 digest
 */
async function hashDocument(file, options = {}) {
    const algorithm = options.algorithm ?? 'sha256';
    const encoding = options.encoding ?? 'hex';
    if (!file) {
        throw new errors_1.ValidationError('File buffer is required');
    }
    if (file.length === 0) {
        throw new errors_1.ValidationError('File buffer cannot be empty');
    }
    if (algorithm === 'blake3') {
        const { createHash } = await importBlake3();
        return createHash().update(file).digest(encoding);
    }
    return (0, crypto_1.createHash)(algorithm).update(file).digest(encoding);
}
/**
 * Compute hash of a file on disk (Node.js only).
 */
async function hashFile(filePath, options = {}) {
    if (typeof globalThis !== 'undefined' && globalThis.window !== undefined) {
        throw new Error('hashFile is only available in Node.js environments.');
    }
    const fs = await Promise.resolve().then(() => __importStar(require('fs')));
    const algorithm = options.algorithm ?? 'sha256';
    const encoding = options.encoding ?? 'hex';
    if (!fs.existsSync(filePath)) {
        throw new errors_1.ValidationError(`File not found: ${filePath}`);
    }
    if (algorithm === 'blake3') {
        const { createHash } = await importBlake3();
        const hasher = createHash();
        const rs = fs.createReadStream(filePath);
        return new Promise((resolve, reject) => {
            rs.on('error', reject);
            rs.on('data', chunk => hasher.update(chunk));
            rs.on('end', () => resolve(hasher.digest(encoding)));
        });
    }
    return new Promise((resolve, reject) => {
        const hash = (0, crypto_1.createHash)(algorithm);
        const rs = fs.createReadStream(filePath);
        rs.on('error', reject);
        rs.on('data', chunk => hash.update(chunk));
        rs.on('end', () => resolve(hash.digest(encoding)));
    });
}
/**
 * Hash a file using streaming for large files.
 */
async function hashFileStream(filePath, options = {}) {
    return hashFile(filePath, options);
}
/**
 * Hash with progress reporting for large files.
 */
async function hashFileWithProgress(filePath, onProgress, options = {}) {
    if (typeof globalThis !== 'undefined' && globalThis.window !== undefined) {
        throw new Error('hashFileWithProgress is only available in Node.js environments.');
    }
    const { statSync, createReadStream } = await Promise.resolve().then(() => __importStar(require('fs')));
    const algorithm = options.algorithm ?? 'sha256';
    const encoding = options.encoding ?? 'hex';
    const totalBytes = statSync(filePath).size;
    let bytesProcessed = 0;
    let hasher;
    if (algorithm === 'blake3') {
        const { createHash } = await importBlake3();
        hasher = createHash();
    }
    else {
        hasher = (0, crypto_1.createHash)(algorithm);
    }
    return new Promise((resolve, reject) => {
        const stream = createReadStream(filePath, {
            highWaterMark: 64 * 1024
        });
        stream.on('data', (chunk) => {
            hasher.update(chunk);
            bytesProcessed += chunk.length;
            onProgress?.(bytesProcessed, totalBytes);
        });
        stream.on('end', () => resolve(hasher.digest(encoding)));
        stream.on('error', reject);
    });
}
/**
 * Compute hash of a base64 string.
 */
async function hashBase64(base64, options = {}) {
    if (!base64 || typeof base64 !== 'string') {
        throw new errors_1.ValidationError('Valid base64 string is required');
    }
    const buffer = Buffer.from(base64, 'base64');
    return hashDocument(buffer, options);
}
/**
 * Compute hash from a Readable stream.
 */
async function hashStream(stream, options = {}) {
    if (!stream)
        throw new errors_1.ValidationError('Stream is required');
    const algorithm = options.algorithm ?? 'sha256';
    const encoding = options.encoding ?? 'hex';
    let hasher;
    if (algorithm === 'blake3') {
        const { createHash } = await importBlake3();
        hasher = createHash();
    }
    else {
        hasher = (0, crypto_1.createHash)(algorithm);
    }
    // Node.js stream
    if (typeof stream.on === 'function') {
        return new Promise((resolve, reject) => {
            stream.on('error', reject);
            stream.on('data', (chunk) => hasher.update(chunk));
            stream.on('end', () => resolve(hasher.digest(encoding)));
        });
    }
    // Web stream
    if (typeof stream.getReader === 'function') {
        const reader = stream.getReader();
        while (true) {
            const { done, value } = await reader.read();
            if (done)
                break;
            hasher.update(value);
        }
        return hasher.digest(encoding);
    }
    throw new errors_1.ValidationError('Unsupported stream type provided to hashStream');
}
/**
 * Validate that a string is a valid hex digest for the given algorithm.
 */
function isValidHash(hash, algorithm = 'sha256') {
    if (typeof hash !== 'string')
        return false;
    const info = getAlgorithmInfo(algorithm);
    const regex = new RegExp(`^[a-f0-9]{${info.outputLength}}$`);
    return regex.test(hash);
}
/**
 * Normalize a hash string.
 */
function normalizeHash(hash) {
    return hash.trim().toLowerCase();
}
