// @sipheron/vdr-core/dist/hash/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.truncateHash = exports.formatFileSize = exports.constantTimeCompare = exports.bufferToHex = exports.hexToBuffer = exports.hashAuto = exports.hashDocumentBrowser = exports.getAlgorithmInfo = exports.normalizeHash = exports.isValidHash = exports.hashBase64 = exports.hashStream = exports.hashFileWithProgress = exports.hashFileStream = exports.hashFile = exports.hashDocument = void 0;
/**
 * @module hash
 *
 * @description
 * Public barrel for the `hash` sub-system.
 *
 * Exports all hashing primitives and low-level buffer utilities from a single
 * import path. Consumers of `@sipheron/vdr-core` should import via the root
 * entry point; internal modules import from `'../hash'`.
 *
 * ### Exported functions
 * From `hash/sha256.ts` (cryptographic core):
 * - `hashDocument` — SHA-256 of a `Buffer`
 * - `hashFile`     — SHA-256 of a file on disk (Node.js only)
 * - `hashStream`   — SHA-256 of a readable stream
 * - `hashBase64`   — SHA-256 of a base64-encoded string
 * - `isValidHash`  — Validates 64-char hex format
 * - `normalizeHash`— Lowercases and trims a hash string
 *
 * From `hash/utils.ts` (low-level helpers):
 * - `hexToBuffer`, `bufferToHex`, `constantTimeCompare`,
 *   `formatFileSize`, `truncateHash`
 */
var algorithms_1 = require("./algorithms");
Object.defineProperty(exports, "hashDocument", { enumerable: true, get: function () { return algorithms_1.hashDocument; } });
Object.defineProperty(exports, "hashFile", { enumerable: true, get: function () { return algorithms_1.hashFile; } });
Object.defineProperty(exports, "hashFileStream", { enumerable: true, get: function () { return algorithms_1.hashFileStream; } });
Object.defineProperty(exports, "hashFileWithProgress", { enumerable: true, get: function () { return algorithms_1.hashFileWithProgress; } });
Object.defineProperty(exports, "hashStream", { enumerable: true, get: function () { return algorithms_1.hashStream; } });
Object.defineProperty(exports, "hashBase64", { enumerable: true, get: function () { return algorithms_1.hashBase64; } });
Object.defineProperty(exports, "isValidHash", { enumerable: true, get: function () { return algorithms_1.isValidHash; } });
Object.defineProperty(exports, "normalizeHash", { enumerable: true, get: function () { return algorithms_1.normalizeHash; } });
Object.defineProperty(exports, "getAlgorithmInfo", { enumerable: true, get: function () { return algorithms_1.getAlgorithmInfo; } });
var browser_1 = require("./browser");
Object.defineProperty(exports, "hashDocumentBrowser", { enumerable: true, get: function () { return browser_1.hashDocumentBrowser; } });
Object.defineProperty(exports, "hashAuto", { enumerable: true, get: function () { return browser_1.hashAuto; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "hexToBuffer", { enumerable: true, get: function () { return utils_1.hexToBuffer; } });
Object.defineProperty(exports, "bufferToHex", { enumerable: true, get: function () { return utils_1.bufferToHex; } });
Object.defineProperty(exports, "constantTimeCompare", { enumerable: true, get: function () { return utils_1.constantTimeCompare; } });
Object.defineProperty(exports, "formatFileSize", { enumerable: true, get: function () { return utils_1.formatFileSize; } });
Object.defineProperty(exports, "truncateHash", { enumerable: true, get: function () { return utils_1.truncateHash; } });
