// @sipheron/vdr-core/dist/hash/sha256.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.hashDocument = hashDocument;
exports.hashFile = hashFile;
exports.hashFileStream = hashFileStream;
exports.hashFileWithProgress = hashFileWithProgress;
exports.hashBase64 = hashBase64;
exports.hashStream = hashStream;
exports.isValidHash = isValidHash;
exports.normalizeHash = normalizeHash;
/**
 * @module hash/sha256
 *
 * @description
 * SHA-256 hashing utilities — the cryptographic foundation of the VDR SDK.
 *
 * All hashing runs **100 % locally** using Node.js's built-in `crypto` module.
 * Document bytes are never transmitted to any server; only the resulting
 * 64-character hex digest is used for anchoring and verification.
 *
 * ## Functions
 * | Function            | Environment    | Input            |
 * |---------------------|----------------|------------------|
 * | `hashDocument(buf)` | Node + Browser | `Buffer`         |
 * | `hashFile(path)`    | Node only    n | filesystem path  |
 * | `hashStream(stream)`| Node + Browser | readable stream  |
 * | `hashBase64(b64)`   | Node + Browser | Base64 string    |
 * | `isValidHash(h)`    | —              | any string       |
 * | `normalizeHash(h)`  | —              | any string       |
 *
 * ## Security properties
 * - **Non-reversible** — SHA-256 is a one-way function; the document cannot
 *   be recovered from its hash.
 * - **Collision-resistant** — probability of two different documents sharing
 *   the same hash is negligible (2⁻²⁵⁶).
 * - **Deterministic** — the same document always produces the same hash.
 *
 * @example
 * ```ts
 * import { hashDocument, hashFile } from '@sipheron/vdr-core'
 * import fs from 'fs'
 *
 * // From a Buffer (browser or Node)
 * const hash1 = await hashDocument(fs.readFileSync('./contract.pdf'))
 *
 * // From a file path (Node only)
 * const hash2 = await hashFile('./contract.pdf')
 *
 * console.log(hash1 === hash2) // true
 * ```
 */
const crypto_1 = require("crypto");
const errors_1 = require("../errors");
/**
 * Compute SHA-256 hash of a document buffer.
 *
 * This function always runs locally — the document bytes
 * are never transmitted anywhere. Only the resulting
 * 64-character hex digest is used for anchoring.
 *
 * @param file - Document as Buffer
 * @returns 64-character lowercase hex string
 *
 * @example
 * const hash = await hashDocument(fs.readFileSync('./contract.pdf'))
 * // a3f4b2c1d8e9f0a1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5
 */
async function hashDocument(file) {
    if (!file) {
        throw new errors_1.ValidationError('File buffer is required');
    }
    if (file.length === 0) {
        throw new errors_1.ValidationError('File buffer cannot be empty');
    }
    return (0, crypto_1.createHash)('sha256').update(file).digest('hex');
}
/**
 * Compute SHA-256 hash of a file on disk (Node.js only).
 *
 * @param filePath - Path to the file
 * @returns 64-character lowercase hex string
 */
async function hashFile(filePath) {
    if (typeof globalThis !== 'undefined' && globalThis.window !== undefined) {
        throw new Error('hashFile is only available in Node.js environments.');
    }
    const fs = await Promise.resolve().then(() => __importStar(require('fs')));
    if (!fs.existsSync(filePath)) {
        throw new errors_1.ValidationError(`File not found: ${filePath}`);
    }
    return new Promise((resolve, reject) => {
        const hash = (0, crypto_1.createHash)('sha256');
        const rs = fs.createReadStream(filePath);
        rs.on('error', reject);
        rs.on('data', chunk => hash.update(chunk));
        rs.on('end', () => resolve(hash.digest('hex')));
    });
}
/**
 * Hash a file using streaming — handles files of any size
 * without loading entire file into memory.
 * Memory usage: O(chunk size) not O(file size)
 */
async function hashFileStream(filePath) {
    if (typeof globalThis !== 'undefined' && globalThis.window !== undefined) {
        throw new Error('hashFileStream is only available in Node.js environments.');
    }
    const { createReadStream } = await Promise.resolve().then(() => __importStar(require('fs')));
    return new Promise((resolve, reject) => {
        const hash = (0, crypto_1.createHash)('sha256');
        const stream = createReadStream(filePath, {
            highWaterMark: 64 * 1024 // 64KB chunks
        });
        stream.on('data', (chunk) => hash.update(chunk));
        stream.on('end', () => resolve(hash.digest('hex')));
        stream.on('error', reject);
    });
}
/**
 * Hash with progress reporting for large files.
 * onProgress receives bytes processed and total bytes.
 */
async function hashFileWithProgress(filePath, onProgress) {
    if (typeof globalThis !== 'undefined' && globalThis.window !== undefined) {
        throw new Error('hashFileWithProgress is only available in Node.js environments.');
    }
    const { statSync, createReadStream } = await Promise.resolve().then(() => __importStar(require('fs')));
    const totalBytes = statSync(filePath).size;
    let bytesProcessed = 0;
    return new Promise((resolve, reject) => {
        const hash = (0, crypto_1.createHash)('sha256');
        const stream = createReadStream(filePath, {
            highWaterMark: 64 * 1024
        });
        stream.on('data', (chunk) => {
            hash.update(chunk);
            bytesProcessed += chunk.length;
            onProgress?.(bytesProcessed, totalBytes);
        });
        stream.on('end', () => resolve(hash.digest('hex')));
        stream.on('error', reject);
    });
}
/**
 * Compute SHA-256 hash of a string to handle Base64 properly.
 *
 * @param base64 - Base64 encoded document string
 * @returns 64-character lowercase hex string
 */
async function hashBase64(base64) {
    if (!base64 || typeof base64 !== 'string') {
        throw new errors_1.ValidationError('Valid base64 string is required');
    }
    const buffer = Buffer.from(base64, 'base64');
    return hashDocument(buffer);
}
/**
 * Compute SHA-256 hash from a Readable stream.
 * Compatible with Node.js streams or Web Streams.
 *
 * @param stream - A readable stream
 * @returns 64-character lowercase hex string
 */
async function hashStream(stream) {
    if (!stream)
        throw new errors_1.ValidationError('Stream is required');
    // Node.js stream
    if (typeof stream.on === 'function') {
        return new Promise((resolve, reject) => {
            const hash = (0, crypto_1.createHash)('sha256');
            stream.on('error', reject);
            stream.on('data', (chunk) => hash.update(chunk));
            stream.on('end', () => resolve(hash.digest('hex')));
        });
    }
    // Web stream
    if (typeof stream.getReader === 'function') {
        const reader = stream.getReader();
        const hash = (0, crypto_1.createHash)('sha256');
        while (true) {
            const { done, value } = await reader.read();
            if (done)
                break;
            hash.update(value);
        }
        return hash.digest('hex');
    }
    throw new errors_1.ValidationError('Unsupported stream type provided to hashStream');
}
/**
 * Validate that a string is a valid SHA-256 hex digest.
 *
 * @param hash - String to validate
 * @returns true if valid SHA-256 hex string
 */
function isValidHash(hash) {
    return typeof hash === 'string' && /^[a-f0-9]{64}$/.test(hash);
}
/**
 * Normalize a hash string to lowercase.
 * SHA-256 output is case-insensitive — normalize to lowercase
 * before any comparison.
 */
function normalizeHash(hash) {
    return hash.trim().toLowerCase();
}
